rfcs/007 - Peer review infra changes.md

```diff
@@ -0,0 +1,125 @@
+- Allow plain SQL or have an introspective ORM
+- Free and ideally open source
+- Baseline/ignore certain migrations
+- Be resilient to `iasql_ugprade`
+
+Data migrations require being able to baseline migrations to avoid duplicate record violations on `iasql_upgrade`. Flyway comes closest to meeting all our requirements. However, it is not possible to baseline/ignore existing migrations on their free product.
+
+We could configure `iasql_upgrade` to preserve a `migrations` table and make the name of the table configurable, but the baselining issue would remain as `iasql_upgrade` can change the schema which renders existing migrations on past IaSQL versions invalid. As a result, we need an opinionated solution to unblock peer review of infra changes.
+
+## Proposal
+
+Take inspiration from [RFC 006](./006%20-%20Replicate%20changes%20between%20staging%20and%20prod%20RFC.md) and create a how-to guide in the documentation much like the guide on [how to replicate changes across infra environments](https://iasql.com/docs/envs/) along with an example repository that walks users through how to code review their changes using **convention over configuration** such that the workflow:
+- works with any cloud identity and access management structure
+- works with `iasql_upgrade`
+- works with any source control solution without the need for special access to it
+- preserves the ability to `iasql_preview` changes in development and get the type-safety feedback from the dashboard editor
+
+The workflow simply suggests users run `iasql_begin`, perform all the necessary changes, and before doing `iasql_commit` they manually put the output of `iasql_preview` and the SQL generated thus far into a markdown file within a repository of their liking. Once the PR is approved by their peers and landed, they can manually run `iasql_commit`. We can require a parameter for `iasql_commit(message)` which is a string attached as metadata for the transaction that can be set to be the PR URL. Best practices are thus enforced using convention and not configuration with a complete reference of what was done via the IaSQL audit log if necessary.
+
+The required work in the IaSQL engine is associating entries in the audit log with specific transactions and adding a `get_sql_for_transaction()` that returns the SQL for the current transaction so users don't have to provide a timestamp for `get_sql_since()`. Additionally, we can modify transactions so they are active past 30 minutes and do not expire at a given point in time as users are making changes to the database.
+
+Down the line, we can revisit another alternative for a longer-term implementation and/or automate different parts of the convention for this workflow when we have more data points and resources. It might be as simple as adding a job template in each CI platform that only requires a Postgres connection string and can be triggered manually within the CI/CD to open a PR with the details of the current transaction.
```

Comment:
If we are going with "convention over configuration" it would be both 1) pretty easy to do (a pure pl/pgsql function if desired) and 2) more "conventional" if there was an `iasql_create_review(title text, description text)` function they call that internally creates the context of the markdown file consisting of the title and description of the change, followed by ` ```sql ` and `<table>` formatted blocks powered by `get_sql_for_transaction()` and `iasql_preview()`

Then they just have to copy-paste this into github or whatever version control system their team uses.

The documentation can describe what it's doing under the hood with these two functions but reduce the friction involved in formatting the output for consumption.

Reply:
If the function has no side effect and just outputs a bunch of text that sounds great!

Reply:
It should have zero side-effects (excluding the `cloud.read` calls that `iasql_preview` already would trigger, but those should not ever change anything in the cloud), yes